### PR TITLE
Propagate parent and providers for go SDK calls

### PR DESCRIPTION
**Pull request by EvanBoyle**
Fixes https://github.com/pulumi/pulumi/issues/3494
I added support to propagate a parent in `Invoke` calls. If specified, parent is used as a backstop to lookup providers for the operation. 

While I was in this space, I also did the same for `RegisterResource`, and `ReadResource`. 

`ReadResource` has a bug in the API that is blocking me from writing the proper tests https://github.com/pulumi/pulumi/issues/3562. I'll be addressing that bug and adding the appropriate tests in another PR. 

I had to update the version of `assert` in order to use `assert.Same` to verify pointer equality.

### CODE REVIEW

File: sdk/go/pulumi/context.go [hunk 1]
```diff
@@ -323,6 +345,67 @@ func (state *ResourceState) ID() IDOutput {
...
+			provider = opt.Provider
+		}
+		if len(providers) == 0 && opt.Providers != nil {
+			for k, v := range opt.Providers {
+				providers[k] = v
+			}
+		}
+	}
+
+	mergedProviders := make(map[string]ProviderResource)
```
Comment (pgavlin) [inline sdk/go/pulumi/context.go]: You can eliminate this by doing the following below:

```
 	// copy parent providers
 	if parent != nil {
 		rs, ok := parent.(*ResourceState)
 		if ok {
 			for k, v := range rs.providers {
 				if _, has := providers[k]; !has {
 					providers[k] = v
 				}
 			}
 		}
  	}
```

File: sdk/go/pulumi/context.go [hunk 2]
```diff
@@ -255,6 +269,12 @@ func (ctx *Context) RegisterResource(
 	// Create resolvers for the resource's outputs.
 	res := makeResourceState(custom, props)
 
+	providers := mergeProviders(t, opts...)
+
+	for k, v := range providers {
+		res.providers[k] = v
```
Comment (pgavlin) [inline sdk/go/pulumi/context.go]: No need to copy--might as well just slam the returned map into `res`.

File: sdk/go/pulumi/context.go [hunk 3]
```diff
@@ -531,6 +617,13 @@ func (ctx *Context) getOpts(opts ...ResourceOpt) (URN, []URN, bool, string, bool
 		}
 	}
 
+	providers := mergeProviders(t, opts...)
```
Comment (pgavlin) [inline sdk/go/pulumi/context.go]: Feels like we should just pull this off of `*ResourceState` rather than recomputing it.

Comment (EvanBoyle) [reply sdk/go/pulumi/context.go]: Done. Thanks for pointing this out. I imagine the concern here is perf related, as piping through the additional parameter through two functions is a little ugly? Looking more closely, we're computing a set of sync and async properties, and we should probably have `prepareOpts` functions for each to make this a little cleaner.

Comment (EvanBoyle) [reply sdk/go/pulumi/context.go]: (It's been a minute since I had to worry at object allocation level perf)

**Pull request merged**